dwave/system/temperatures.py

```diff
@@ -0,0 +1,542 @@
+       problem (random couplers J, zero external field h).
+       Establish a temperature estimate by maximum pseudo-likelihood. 
+
+       Note that due to the complicated freeze-out properties of hard models,
+       such as large scale spin-glasses, deviation from a classical Boltzmann 
+       distribution is anticipated.
+       Nevertheless, the T estimate can always be interpreted as an estimator
+       of local excitations rates. For example T will be 0 if only 
+       local minima are returned (even if some of the local minima are not 
+       ground states).
+       
+       >>> import dimod
+       >>> from dwave.system.temperatures import maximum_pseudolikelihood_temperature
+       >>> from dwave.system import DWaveSampler
+       >>> from random import random
+       >>> sampler = DWaveSampler() 
+       >>> bqm = dimod.BinaryQuadraticModel.from_ising({}, {e : 1-2*random() for e in sampler.edgelist})
+       >>> sampleset = sampler.sample(bqm, num_reads=100, auto_scale=False)
+       >>> T,T_bootstrap =  maximum_pseudolikelihood_temperature(bqm,sampleset) 
+       >>> print('Effective temperature ',T)    # doctest: +SKIP
+       
+    See also:
+
+        https://doi.org/10.3389/fict.2016.00023
+
+        https://www.jstor.org/stable/25464568
+        
+    '''
+    
+    T_estimate = 0
+    T_bootstrap_estimates = np.zeros(bootstrap_size)
+
+    #Check for largest local excitation in every sample, and over all samples
+    if site_energy is None:
+        if bqm == None or sampleset == None:
+            raise ValueError('site_energy can only be derived if both'
+                             'bqm and sampleset are provided as arguments')
+        site_energy = effective_field(bqm,
+                                      sampleset,
+                                      current_state_energy = True)
+        
+    max_excitation = np.max(site_energy,axis=1)
+    max_excitation_all =  np.max(max_excitation)
+    if max_excitation_all <= 0:
+        #There are no local excitations present in the sample set, therefore
+        #the temperature is estimated as 0. 
+        pass
+    else:
+        
+        def d_mean_log_pseudo_likelihood(x):
+            #Derivative of mean (w.r.t samples) log pseudo liklihood amounts
+            #to local energy matching criteria
+            #O = sum_i sum_s e_i(s) P(s,i) #s = sample, i = variable index
+            #e_i(s) is energy lost in flipping spin s_i against current assignment.
+            #P(s,i) = 1/(1 + exp[x e_i(s)]), probability to flip against current state.
+            #x = -1/T
+            expFactor = np.exp(site_energy*x)
+            return np.sum(site_energy/(1 + expFactor))
+        
+        def dd_mean_log_pseudo_likelihood(x):
+            #For very large or small site energies this may be numerically
+            #unstable, therefore we prefer a bisection section.
+            expFactor = np.exp(site_energy*x)
```

Comment:
When running the example of lines 213 - 220, I keep getting a warning:

```
C:\Users\jpasvolsky\!git_ADTT\system\dwave-system\dwave\system\temperatures.py:259: RuntimeWarning: overflow encountered in exp
  return np.sum(site_energy/(1 + expFactor))
```

Reply:
It expFactor evaluates to infinity due to the overflow, that doesn't cause an issue finite/infinity is still 0

Reply:
But you might want to put that in a `try:` statement to not print out the warning

Reply:
I think I'll filter out this specific error by default, given that it is understood not to be problematic: https://queirozf.com/entries/suppressing-ignoring-warnings-in-python-reference-and-examples